## 对作业题的第9(a)题，使用scipy的linprog进行单纯形法的线性规划

In [1]:
from scipy import optimize as op
import numpy as np

c=np.array([6,14,13])    # 目标函数
A_ub=np.array([[1,4,2],[1,2,4]])  # 不等式约束
B_ub=np.array([48,60])
A_eq=np.array([[-1,1,1]])   # 等式约束
B_eq=np.array([4])
# x1,x2,x3的约束条件默认都非负
res=op.linprog(-c,A_ub,B_ub,method='simplex')    # 目标函数为Min
print(res)

     con: array([], dtype=float64)
     fun: -294.0
 message: 'Optimization terminated successfully.'
     nit: 3
   slack: array([0., 0.])
  status: 0
 success: True
       x: array([36.,  0.,  6.])


`op.linprog`是求目标函数的最小值的，而原问题要求最大值，因此第一个参数里面加一个负号，得到最优解为$ x=(36,0,6)^T $时最小值为$ -294 $，因此原问题的最大值为$294$

----
## 求解第4题的(a)

In [3]:
from scipy import optimize as op
import numpy as np

c=np.array([2,-1,2])    # 目标函数
A_ub=np.array([[-1,1,-1]])  # 不等式约束
B_ub=np.array([6])
A_eq=np.array([[-1,1,1]])   # 等式约束
B_eq=np.array([4])
# x1,x2,x3的约束条件
x1=(None,0) # x1<=0
x2=(0,None) # x2>=0
x3=(None,None)  # x3无约束
res=op.linprog(c,A_ub,B_ub,A_eq,B_eq,bounds=(x1,x2,x3),method='simplex')    # 目标函数为Min，约束条件bounds
print(res)

     con: array([0.])
     fun: -12.0
 message: 'Optimization terminated successfully.'
     nit: 2
   slack: array([0.])
  status: 0
 success: True
       x: array([-5.,  0., -1.])


根据返回值，得到最优解为$x=(-5,0,-1)^T$， 此时最小值为$-12$

------

## 第4题(b)

In [1]:
from scipy import optimize as op
import numpy as np

c=np.array([2,1,3,1])    # 目标函数
A_ub=np.array([[1,1,1,1],[-1,0,2,-2]])  # 不等式约束
B_ub=np.array([7,-1])
A_eq=np.array([[2,-3,5,0]])   # 等式约束
B_eq=np.array([-8])
# x1,x2,x3的约束条件
x1=(0,None) # x1>=0
x2=(None,0) # x2<=0
x3=(0,None)  # x4<=0
x4=(None,None)
res=op.linprog(c,A_ub,B_ub,A_eq,B_eq,bounds=(x1,x2,x3,x4),method='simplex')    # 目标函数为Min
print(res)

     con: array([-8.])
     fun: 0.5
 message: "Phase 1 of the simplex method failed to find a feasible solution. The pseudo-objective function evaluates to 8.0e+00 which exceeds the required tolerance of 1e-09 for a solution to be considered 'close enough' to zero to be a basic solution. Consider increasing the tolerance to be greater than 8.0e+00. If this tolerance is unacceptably  large the problem may be infeasible."
     nit: 2
   slack: array([6.5, 0. ])
  status: 2
 success: False
       x: array([0. , 0. , 0. , 0.5])


使用单纯性法规划失败，无最优解